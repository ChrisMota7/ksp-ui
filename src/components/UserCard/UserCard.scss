.user-card {
    width: 300px;
    border-radius: 10px;
    padding: 10px;
    margin: 10px;
    -webkit-box-shadow: 0px 0px 8px -4px rgba(0,0,0,0.79);
    -moz-box-shadow: 0px 0px 8px -4px rgba(0,0,0,0.79);
    box-shadow: 0px 0px 8px -4px rgba(0,0,0,0.79);

    &__fullname {
        font-weight: bold;
    }

    &__typeOfUser {
        color: gray;
    }

    &__delete {
        width: 100%;
        display: flex;
        justify-content: flex-end;
    }
}