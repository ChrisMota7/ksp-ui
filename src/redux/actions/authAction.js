import { get, post } from "@/utils/api"
import { jwtDecode } from "jwt-decode";
import { AUTHENTICATE_USER } from "../reducers/authReducer";

export const login = (email, password) => async (dispatch) => {
    try {
        console.log("here")
        const { access } = await post("/api/token/", {
            email,
            password
        })
        
        const { user_id } = jwtDecode(access)

        const { is_admin } = await get("/auth/user-type/", {
            Authorization: `Bearer ${access}`,
        })
        
        localStorage.setItem("jwt", access)
        localStorage.setItem("userid", user_id)
        localStorage.setItem("user_email", email)
        localStorage.setItem("isAdmin", is_admin)

        dispatch({
            type: AUTHENTICATE_USER,
            payload: {
                jwt: access,
                userid: user_id,
                isAdmin: is_admin,
            }
        })

        console.log("access",access)
        console.log("user_id",user_id)
        console.log("isAdmin",is_admin)

        return { userAuthenticationSuccessfully: true }
    } catch (e) {
        console.log("error", e)

        return { userAuthenticationSuccessfully: false }
    } 
}