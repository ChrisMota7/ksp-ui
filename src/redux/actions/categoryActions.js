import { get, post, put } from "@/utils/api"
import { SET_CATEGORIES, SET_DASHBOARD, ADD_PROBLEM, SET_CATEGORIES_ALL} from "../reducers/categoryReducer"

export const getCategoriesAll = () => async (dispatch) => {
    try {
        const categories = await get('/helpdesk/categorias/')

        dispatch({
            type: SET_CATEGORIES_ALL,
            payload: categories
        })

        return { setCategoryTableSuccessfully: true }
    } catch (e) {
        console.log("error", e)

        return { setCategoryTableSuccessfully: false }
    }
}

export const getProblems = () => async (dispatch) => {
    try {
        const categories = await get('/helpdesk/problemas/')

        console.log("categories",categories)

        dispatch({
            type: SET_CATEGORIES,
            payload: categories
        })

        return { setCategoryTableSuccessfully: true }
    } catch (e) {
        console.log("error", e)

        return { setCategoryTableSuccessfully: false }
    }
}

export const createProblem = (name, categoria_id, prioridad_id) => async (dispatch) => {

    const body = new FormData()
    body.append('name', name)
    body.append('categoria', categoria_id)
    body.append('prioridad', prioridad_id)

    try {
        const response = await post("/helpdesk/problemas/", body, {
        })

        console.log("response",response)

        return { problemCreatedSuccessfully: true }
    } catch (e) {
        console.log("error", e)

        return { problemCreatedSuccessfully: false }
    }
}

export const createCategorie = (name) => async (dispatch) => {

    const body = new FormData()
    body.append('name', name)

    try {
        const response = await post("/helpdesk/categorias/crear/", body, {
        })

        console.log("response",response)

        return { categorieCreatedSuccessfully: true }
    } catch (e) {
        console.log("error", e)

        return { categorieCreatedSuccessfully: false }
    }
}

export const updateProblem = (categoryId, categoryName) => async (dispatch) => {
    try {
        const response = await put(`/helpdesk/problems/${categoryId}/update/`, 
        { 
            name: categoryName 
        })

        return { setUpdateCategerySuccess: true }
    } catch (e) {
        console.log("error", e)

        return { setUpdateCategerySuccess: false }
    }
}

export const deleteProblem = (categoryId) => async (dispatch) => {
    try {
        await put(`/helpdesk/problems/${categoryId}/delete/`);

        return { setDeleteCategerySuccess: true }
    } catch (e) {
        console.log("error", e)

        return { setDeleteCategerySuccess: false }
    }
}

export const getDashboard = () => async (dispatch) => {
    try {
        const dashboard = await get("/helpdesk/stats/tickets/");

        console.log("datos dashboard", dashboard)
        dispatch({
            type: SET_DASHBOARD,
            payload: dashboard
        })
        return { setDashboardSuccessfully: true }
    } catch (e) {
        console.log("error", e)

        return { setDashboardSuccessfully: false }
    }
}

export const updateCategories = (categoryId, categoryName) => async (dispatch) => {
    try {
        const response = await put(`/helpdesk/categories/${categoryId}/update/`, 
        { 
            name: categoryName 
        })
        return { setUpdateCategerySuccess: true }
    } catch (e) {
        console.log("error", e)
        return { setUpdateCategerySuccess: false }
    }
}

export const deleteCategories = (categoryId) => async (dispatch) => {
    try {
        await put(`/helpdesk/categories/${categoryId}/delete/`)
        return { setDeleteCategerySuccess: true }
    } catch (e) {
        console.log("error", e)
        return { setDeleteCategerySuccess: false }
    }
}