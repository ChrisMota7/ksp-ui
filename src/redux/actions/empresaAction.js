import { get, post, put } from "@/utils/api";
import { SET_EMPRESAS, SET_EMPRESAS_ALL } from "../reducers/empresaReducer";

export const getEmpresasAll = () => async (dispatch) => {
    try {
        const empresas = await get('/user/empresas/');
        dispatch({
            type: SET_EMPRESAS_ALL,
            payload: empresas
        });
        return { setEmpresaTableSuccessfully: true };
    } catch (e) {
        console.log("error", e);
        return { setEmpresaTableSuccessfully: false };
    }
}

export const getEmpresasTable = () => async (dispatch) => {
    try {
        const empresas = await get('/user/empresas/table/');
        console.log("AcciÃ³n getEmpresasTable - empresas:", empresas);
        dispatch({
            type: SET_EMPRESAS,
            payload: empresas,
        });
        return { setEmpresaTableSuccessfully: true };
    } catch (e) {
        console.log("error", e);
        return { setEmpresaTableSuccessfully: false };
    }
};

export const createEmpresa = (nombre) => async (dispatch) => {
    const body = new FormData();
    body.append('nombre', nombre);

    try {
        const response = await post("/user/empresas/", body);
        return { empresaCreatedSuccessfully: true };
    } catch (e) {
        console.log("error", e);
        return { empresaCreatedSuccessfully: false };
    }
}

export const updateEmpresa = (empresaId, data) => async (dispatch) => {
    try {
        const response = await put(`/user/empresas/${empresaId}/`, data);
        return { setUpdateEmpresaSuccess: true };
    } catch (e) {
        console.log("error", e);
        return { setUpdateEmpresaSuccess: false };
    }
}

export const deleteEmpresa = (empresaId) => async (dispatch) => {
    try {
        await put(`/user/empresas/${empresaId}/delete/`);
        return { setDeleteEmpresaSuccess: true };
    } catch (e) {
        console.log("error", e);
        return { setDeleteEmpresaSuccess: false };
    }
}
