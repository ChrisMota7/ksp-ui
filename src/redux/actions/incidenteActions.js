import { get, post, put } from "@/utils/api";
import { SET_TIPO_INCIDENTES, SET_INCIDENTES, SET_INCIDENTE_INFO, selectIncidenteInfo } from "../reducers/incidenteReducer";

export const getTipoIncidentes = () => async (dispatch) => {
    try {
        const tiposIncidentes = await get('/helpdesk/tipos-incidente/');
        dispatch({ type: SET_TIPO_INCIDENTES, payload: tiposIncidentes });
        return { setTipoIncidentesSuccessfully: true };
    } catch (e) {
        console.log("error", e);
        return { setTipoIncidentesSuccessfully: false };
    }
};

export const createTipoIncident = (name, prioridadId) => async (dispatch) => {
    const body = { name, prioridad: prioridadId };
    try {
        const response = await post("/helpdesk/tipos-incidente/", body);
        return { tipoIncidentCreatedSuccessfully: true };
    } catch (e) {
        console.log("error", e);
        return { tipoIncidentCreatedSuccessfully: false };
    }
};

export const updateTipoIncidente = (tipo_incidente_id, data) => async (dispatch) => {
    try {
        const response = await put(`/helpdesk/tipos-incidente/${tipo_incidente_id}/update/`, data);
        return { setUpdateTipoIncidenteSuccess: true };
    } catch (e) {
        console.log("error", e);
        return { setUpdateTipoIncidenteSuccess: false };
    }
};

export const deleteTipoIncidente = (id) => async (dispatch) => {
    try {
        await put(`/helpdesk/tipos-incidente/${id}/delete/`);
        return { setDeleteTipoIncidenteSuccess: true };
    } catch (e) {
        console.log("error", e);
        return { setDeleteTipoIncidenteSuccess: false };
    }
};

export const getIncidentes = () => async (dispatch) => {
    const isAdmin = localStorage.getItem("isAdmin");
    const userId = localStorage.getItem("userid");

    try {
        const incidentes = await get('/helpdesk/incidentes/');

        let filteredIncidentes;
        if (isAdmin === "false") {
            filteredIncidentes = incidentes.filter(incidente => incidente.user.id == userId && incidente.isDeleted === "0");
        } else {
            filteredIncidentes = incidentes.filter(incidente => incidente.isDeleted === "0");
        }

        dispatch({ type: SET_INCIDENTES, payload: filteredIncidentes });
        return { setIncidentesSuccessfully: true };
    } catch (e) {
        console.log("error", e);
        return { setIncidentesSuccessfully: false };
    }
};

export const createIncident = (tipoIncidenteId, puesto, descripcion, acciones, personas, detalles, adicionales, archivos = []) => async (dispatch) => {
    const accessToken = localStorage.getItem("jwt");

    const body = new FormData();
    body.append('tipoIncidente', tipoIncidenteId);
    body.append('puesto', puesto);
    body.append('descripcion', descripcion);
    body.append('acciones', acciones);
    body.append('personas', personas);
    body.append('detalles', detalles);
    body.append('adicionales', adicionales);

    archivos.forEach(archivo => {
        body.append('archivos', archivo); // Asegúrate de que cada archivo se adjunte
    });

    try {
        const response = await post("/helpdesk/incidentes/crear/", body, {
            Authorization: `Bearer ${accessToken}`,
        });

        console.log("response", response);

        const incidentId = response.id;

        return { incidentCreatedSuccessfully: true, newIncidentId: incidentId };
    } catch (e) {
        console.log("error", e);

        return { incidentCreatedSuccessfully: false, newIncidentId: undefined };
    }
};

export const deleteIncidente = (id) => async (dispatch) => {
    try {
        const response = await put(`/helpdesk/incidentes/${id}/delete/`);
        dispatch(getIncidentes());
        return { incidenteDeletedSuccessfully: true };
    } catch (e) {
        console.log("error", e);
        return { incidenteDeletedSuccessfully: false };
    }
};

export const getIncidentInfo = (incidentId) => async (dispatch) => {
    const accessToken = localStorage.getItem("jwt");

    let relatedFiles = [];
    try {
        relatedFiles = await get(`/helpdesk/incidentes/${incidentId}/archivos/`, {
            Authorization: `Bearer ${accessToken}`,
        });
    } catch (e) {
        console.log("Error al obtener archivos:", e);
    }

    try {
        const incidentInfo = await get(`/helpdesk/incidentes/${incidentId}/`, {
            Authorization: `Bearer ${accessToken}`,
        });

        console.log("Información del incidente:", incidentInfo);
        console.log("Archivos relacionados:", relatedFiles);

        dispatch({
            type: SET_INCIDENTE_INFO,
            payload: {
                incidenteInfo: incidentInfo,
                incidentFiles: relatedFiles,
            },
        });

        return { setIncidentInfoSuccessfully: true };
    } catch (e) {
        console.log("Error al obtener información del incidente:", e);

        return { setIncidentInfoSuccessfully: false };
    }
};

export const updateIncident = (incidentId, updatedIncident, archivos = []) => async (dispatch, getState) => {
    try {
        const state = getState();
        const incidente = selectIncidenteInfo(state);

        if (!incidente || !incidente.tipoIncidente) {
            throw new Error("tipoIncidente is undefined");
        }

        const body = new FormData();
        body.append('tipoIncidente', incidente.tipoIncidente);
        body.append('puesto', incidente.puesto);
        body.append('descripcion', incidente.descripcion);
        body.append('acciones', updatedIncident.acciones);
        body.append('personas', updatedIncident.personas);
        body.append('detalles', updatedIncident.detalles);
        body.append('adicionales', updatedIncident.adicionales);

        archivos.forEach(archivo => {
            body.append('archivos', archivo); // Asegúrate de que cada archivo se adjunte
        });

        await put(`/helpdesk/incidentes/${incidentId}/`, body);
        dispatch({ type: 'UPDATE_INCIDENT_SUCCESS' });
        return { incidentUpdatedSuccessfully: true };
    } catch (e) {
        console.error("Error updating incident", e);
        return { incidentUpdatedSuccessfully: false };
    }
};
