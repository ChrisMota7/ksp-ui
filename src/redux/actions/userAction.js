import { get, put, post } from "@/utils/api"
import { SET_FILTERED_USERS, SET_USERS, selectAdminUsers, selectClientUsers, SET_EMPRESAS } from "../reducers/userReducer"
import { filterUsersByFullName } from "@/app/constants"

import { RESET_PASSWORD_REQUEST, RESET_PASSWORD_SUCCESS, RESET_PASSWORD_FAILURE } from "../reducers/userReducer";

export const getEmpresas = () => async (dispatch) => {
    try {
        const response = await get("/user/empresas/");
        console.log('Response from API:', response); 
        dispatch({
            type: SET_EMPRESAS,
            payload: response,
        });
        return response;
    } catch (error) {
        console.error("Error fetching companies", error);
        return [];
    }
};

export const filterUsersByEmpresa = (empresaId) => async (dispatch) => {
    const { getUsersSuccessfully, allUsers } = await dispatch(getUsers());
    if (getUsersSuccessfully && Array.isArray(allUsers)) {
        let filteredUsers;
        if (empresaId === '') {
            filteredUsers = allUsers; // Si `empresaId` está vacío, retorna todos los usuarios
        } else if (empresaId === 'no_empresa') {
            filteredUsers = allUsers.filter(user => !user.empresa); // Filtra usuarios sin empresa
        } else {
            filteredUsers = allUsers.filter(user => user.empresa && user.empresa.id === empresaId);
        }

        dispatch({
            type: SET_FILTERED_USERS,
            payload: filteredUsers
        });
    } else {
        dispatch({
            type: SET_FILTERED_USERS,
            payload: []
        });
    }
};



export const resetPassword = (uid, token, newPassword) => async (dispatch) => {
    dispatch({ type: RESET_PASSWORD_REQUEST });
    try {
        const config = {
            headers: {
                'Content-Type': 'application/json'
            }
        };
        const url = `/user/reset-password/${uid}&/${token}/`;
        const body = { password: newPassword };

        console.log("URL:", url);
        console.log("Request Body:", body);

        const response = await post(url, body, config);

        dispatch({
            type: RESET_PASSWORD_SUCCESS,
            payload: response.data
        });
        console.log("Response Data:", response.data);
        return { resetPasswordSuccessfully: true };
    } catch (error) {
        dispatch({
            type: RESET_PASSWORD_FAILURE,
            payload: error.response && error.response.data.message
                   ? error.response.data.message
                   : error.message
        });
        console.log("Error Response:", error.response);
        return { resetPasswordSuccessfully: false, message: error.message };
    }
};


export const getUsers = (empresaId = null) => async (dispatch) => {
    const accessToken = localStorage.getItem("jwt");
    try {
        let url = "/user/getAllUsers/";
        if (empresaId) {
            url += `?empresa_id=${empresaId}`;
        }
        const allUsers = await get(url, {
            Authorization: `Bearer ${accessToken}`,
        });

        const adminUsers = allUsers.filter((user) => user.isAdmin === "1");
        const clientUsers = allUsers.filter((user) => user.isAdmin === "0");

        dispatch({
            type: SET_USERS,
            payload: {
                adminUsers: adminUsers,
                clientUsers: clientUsers,
            }
        });

        return { getUsersSuccessfully: true, allUsers };
    } catch (e) {
        console.log("error", e);
        return { getUsersSuccessfully: false, allUsers: [] };
    }
};

export const updateUserDetails = (userId, newEmail, newPassword, empresa, newPhoneNumber) => async (dispatch) => {
    const accessToken = localStorage.getItem("jwt");

    try {
        const payload = { email: newEmail, empresa_id: empresa, telefono: newPhoneNumber };
        if (newPassword) {
            payload.password = newPassword;
        }

        const response = await put(`/user/${userId}/UpdateUser/`, payload, {
            Authorization: `Bearer ${accessToken}`
        });

        return { setUpdateUserSuccess: true };
        
    } catch (e) {
        return { setUpdateUserSuccess: false };
    }
}



export const deleteUser = (userId) => async (dispatch) => {
    const accessToken = localStorage.getItem("jwt");

    console.log("accessToken", accessToken);
    console.log("userId", userId);

    try {
        const response = await put(`/user/${userId}/delete/`, {
            Authorization: `Bearer ${accessToken}`,
            userId: userId
        });

        console.log("response", response);
        
        return { setUserDeletedSuccessfully: true };
    } catch (e) {
        console.log("e",e)

        return { setUserDeletedSuccessfully: false };
    }
}

export const filterUsersByName = (userName) => async (dispatch, getState) => {
    const clientUsers = selectClientUsers(getState()).map((message) => message)
    const adminUsers = selectAdminUsers(getState()).map((message) => message)

    const filteredClientUsers = filterUsersByFullName(clientUsers, userName)
    const filteredAdminUsers = filterUsersByFullName(adminUsers, userName)

    const filteredUsers = [...filteredClientUsers, ...filteredAdminUsers]

    dispatch({
        type: SET_FILTERED_USERS,
        payload: filteredUsers
    })
}

export const removeUsersFilter = () => async (dispatch) => {
    dispatch({
        type: SET_FILTERED_USERS,
    })
}

export const sendPasswordResetEmail = (email) => async (dispatch) => {
    const accessToken = localStorage.getItem("jwt");

    try {
        const response = await post(`/user/send-reset-email/`, {
            Authorization: `Bearer ${accessToken}`,
            email: email
        });

        console.log("response", response);
        
        return { sendPasswordResetEmailSuccessfully: true };
        
    } catch (e) {
        console.error("error", e);
        return { sendPasswordResetEmailSuccessfully: false };
    }
}
