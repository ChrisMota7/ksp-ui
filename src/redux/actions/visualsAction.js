import { HIDE_SNACKBAR, SHOW_SNACKBAR } from "../reducers/visualsReducer"

export const showSnackbar = (message, severity) => (dispatch) => {
    dispatch({
        type: SHOW_SNACKBAR,
        payload: {
            message: message,
            severity: severity
        }
    })
}

export const closeSnackbar = () => (dispatch) => {
    dispatch({
        type: HIDE_SNACKBAR
    })
}