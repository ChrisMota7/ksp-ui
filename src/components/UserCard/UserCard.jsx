import { Button } from "@mui/material"
import "./UserCard.scss"
import React, { useState } from 'react'
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogTitle from '@mui/material/DialogTitle';
import { useDispatch } from "react-redux";
import { deleteUser, getUsers } from "@/redux/actions/userAction";
import { showSnackbar } from "@/redux/actions/visualsAction";

const UserCard = ({ userid, name, email, createdAt, isAdmin = false, filteredUser = false, onEdit }) => {
  const dispatch = useDispatch()

  const [openDeleteConfirmation, setOpenDeleteConfirmation] = useState(false)

  const handleCloseDeleteConfirmation = () => {
    setOpenDeleteConfirmation(false)
  }
  
  const handleOpenDeleteConfirmation = () => {
    setOpenDeleteConfirmation(true)
  }

  const handleDelete = async () => {
    const { setUserDeletedSuccessfully } = await dispatch(deleteUser(userid))

    if (setUserDeletedSuccessfully) {
      dispatch(showSnackbar("Usuario eliminado satisfactoriamente", "success"))
      setOpenDeleteConfirmation(false)
      dispatch(getUsers())
    } else {
      dispatch(showSnackbar("Error al eliminar", "error"))
    }
  }

  return (
    <div className='user-card'>
        <p className="user-card__fullname" data-testid="username">{name}</p>
        <p>{email}</p>
        {filteredUser && (
          <p className="user-card__typeOfUser">{isAdmin ? "Administrador" : "Colaborador"}</p>
        )}
        <p>Creación: {new Date(createdAt).toLocaleDateString()}</p>
        <div className="user-card__delete">
            <Button data-testid="delete-button" onClick={handleOpenDeleteConfirmation}>Eliminar</Button>
            <Button data-testid="edit-button" onClick={onEdit}>Editar</Button>
        </div>

        <Dialog
            open={openDeleteConfirmation}
            onClose={handleCloseDeleteConfirmation}
            aria-labelledby="alert-dialog-title"
            aria-describedby="alert-dialog-description"
        >
            <DialogTitle id="alert-dialog-title">
                ¿Desea eliminar a este usuario?
            </DialogTitle>
            <DialogContent>
            <DialogContentText id="alert-dialog-description">
                Al confirmar, se eliminará al usuario {name} y ya no tendrá acceso a la aplicación.
            </DialogContentText>
            </DialogContent>
            <DialogActions>
            <Button data-testid="dialog-cancel-button" onClick={handleCloseDeleteConfirmation} autoFocus>Cancelar</Button>
            <Button data-testid="dialog-delete-button" onClick={handleDelete}>
                Eliminar
            </Button>
            </DialogActions>
        </Dialog>
    </div>
  )
}

export default UserCard