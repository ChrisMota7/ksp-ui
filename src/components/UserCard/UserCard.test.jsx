import { fireEvent, render, waitFor } from '@testing-library/react';
import renderer from "react-test-renderer"
import { useDispatch } from 'react-redux';
import { deleteUser } from '@/redux/actions/userAction';
import { showSnackbar } from "@/redux/actions/visualsAction";

import UserCard from './UserCard';

jest.mock("@/redux/actions/userAction", () => ({
    deleteUser: jest.fn(),
    getUsers: jest.fn(),
}))

jest.mock("@/redux/actions/visualsAction", () => ({
    showSnackbar: jest.fn(),
}))

jest.mock("react-redux", () => ({
    useDispatch: jest.fn()
}))

describe("<UserCard />", () => {
    let dispatchMock

    beforeEach(() => {
        dispatchMock = jest.fn()
        useDispatch.mockReturnValue(dispatchMock)

        jest.clearAllMocks()
    })

    it("should render UserCard correctly", () => {
        const domTree = renderer.create(
            <UserCard 
                userid={1}
                name="Test Name"
                email="test@ksp.com.mx"
                createdAt="2024-05-08T01:18:23.033930Z"
            />
        ).toJSON()
    
        expect(domTree).toMatchSnapshot()
    })

    describe("when delete button is clicked", () => {
        it("should show confirmation modal", () => {
            const { getByTestId, getByText, queryByText } = render(
                <UserCard 
                    userid={1}
                    name="Test Name"
                    email="test@ksp.com.mx"
                    createdAt="2024-05-08T01:18:23.033930Z"
                />
            )

            expect(queryByText("¿Desea eliminar a este usuario?")).not.toBeInTheDocument()

            const deleteButton = getByTestId("delete-button")
            fireEvent.click(deleteButton)

            expect(getByText("¿Desea eliminar a este usuario?")).toBeInTheDocument()
        })

        describe("and cancel button is clicked", () => {
            it("should not delete the user and close the modal", () => {
                const { getByTestId, queryByText, getByText } = render(
                    <UserCard 
                        userid={1}
                        name="Test Name"
                        email="test@ksp.com.mx"
                        createdAt="2024-05-08T01:18:23.033930Z"
                    />
                )

                expect(queryByText("¿Desea eliminar a este usuario?")).not.toBeInTheDocument()

                const deleteButton = getByTestId("delete-button")
                fireEvent.click(deleteButton)
                
                expect(getByText("¿Desea eliminar a este usuario?")).toBeInTheDocument()

                const dialogCancelButton = getByTestId("dialog-cancel-button")
                fireEvent.click(dialogCancelButton)

                expect(deleteUser).not.toBeCalled()

                waitFor(() => {
                    expect(queryByText("¿Desea eliminar a este usuario?")).not.toBeInTheDocument()
                })
            })
        })

        describe("and delete button is clicked", () => {
            it("should delete the user and close the modal", () => {
                dispatchMock.mockReturnValue({
                    setUserDeletedSuccessfully: true
                })

                const { getByTestId, queryByText, getByText } = render(
                    <UserCard 
                        userid={1}
                        name="Test Name"
                        email="test@ksp.com.mx"
                        createdAt="2024-05-08T01:18:23.033930Z"
                    />
                )

                expect(queryByText("¿Desea eliminar a este usuario?")).not.toBeInTheDocument()

                const deleteButton = getByTestId("delete-button")
                fireEvent.click(deleteButton)
                
                expect(getByText("¿Desea eliminar a este usuario?")).toBeInTheDocument()

                const dialogDeleteButton = getByTestId("dialog-delete-button")
                fireEvent.click(dialogDeleteButton)

                expect(deleteUser).toBeCalled()

                waitFor(() => {
                    expect(queryByText("¿Desea eliminar a este usuario?")).not.toBeInTheDocument()
                })
            })
            
            describe("when the user is deleted successfully", () => {
                it("should show snackbar with success message", async () => {
                    dispatchMock.mockReturnValue({
                        setUserDeletedSuccessfully: true
                    })
    
                    const { getByTestId } = render(
                        <UserCard 
                            userid={1}
                            name="Test Name"
                            email="test@ksp.com.mx"
                            createdAt="2024-05-08T01:18:23.033930Z"
                        />
                    )
    
                    const deleteButton = getByTestId("delete-button")
                    await fireEvent.click(deleteButton)
                    
                    const dialogDeleteButton = getByTestId("dialog-delete-button")
                    await fireEvent.click(dialogDeleteButton)
    
                    expect(deleteUser).toBeCalledWith(1)
                    expect(showSnackbar).toBeCalledWith("Usuario eliminado satisfactoriamente", "success")
                })
            })
            
            describe("when the user is NOT deleted successfully", () => {
                it("should show snackbar with success message", async () => {
                    dispatchMock.mockReturnValue({
                        setUserDeletedSuccessfully: false
                    })
    
                    const { getByTestId } = render(
                        <UserCard 
                            userid={1}
                            name="Test Name"
                            email="test@ksp.com.mx"
                            createdAt="2024-05-08T01:18:23.033930Z"
                        />
                    )
    
                    const deleteButton = getByTestId("delete-button")
                    await fireEvent.click(deleteButton)
                    
                    const dialogDeleteButton = getByTestId("dialog-delete-button")
                    await fireEvent.click(dialogDeleteButton)
    
                    expect(deleteUser).toBeCalledWith(1)
                    expect(showSnackbar).toBeCalledWith("Error al eliminar", "error")
                })
            })
        })
    })
})