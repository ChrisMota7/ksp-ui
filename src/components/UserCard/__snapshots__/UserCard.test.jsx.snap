// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserCard /> should render UserCard correctly 1`] = `
<div
  className="user-card"
>
  <p
    className="user-card__fullname"
    data-testid="username"
  >
    Test Name
  </p>
  <p>
    test@ksp.com.mx
  </p>
  <p>
    Creaci√≥n: 
    7/5/2024
  </p>
  <div
    className="user-card__delete"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-testid="delete-button"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Eliminar
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-testid="edit-button"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Editar
    </button>
  </div>
</div>
`;
